File: 46. Permutations/permute.py
from typing import List

class Solution:
  def permute(self, nums: List[int]) -> List[List[int]]:
    permutations = []
    l = len(nums)
    used = [False] * l
    
    def dfs(path):
      if len(path) == l:
        permutations.append(path.copy())
        return
      
      for i in range(l):
        if used[i]:
          continue
        
        used[i] = True
        path.append(nums[i])
        dfs(path)
        path.pop()
        used[i] = False
    
    dfs([])
    return permutations
        
        
nums = [1,2,3]
solution = Solution()
print('permute: ', solution.permute(nums))

File: 46. Permutations/permute.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const permute = (nums) => {
  const permutations = [];
  const l = nums.length;
  const used = Array(l).fill(false);

  const dfs = (path) => {
    if (path.length === l) {
      permutations.push([...path]);
      return;
    }

    for (let i = 0; i < l; i++) {
      if (used[i]) {
        continue;
      }
      used[i] = true;
      path.push(nums[i]);
      dfs(path);
      path.pop();
      used[i] = false;
    }
  };

  dfs([]);
  return permutations;
};


const nums = [1,2,3]
console.log('permute: ', permute(nums));
